### PR TITLE
Better types for immutables

### PR DESCRIPTION
**Pull request by lorem--ipsum**

### CODE REVIEW

**Comment by davidagee** [comment]
We've previously run into a number of runtime issues when tightening up type safety in the application, e.g. default values for inputs no longer being populated or controls crashing the application with certain prop permutations. This is a global change to a base class used widely across the application.

How are we sure that this isn't going to cause unintended consequences?

File: src/base-immutable/base-immutable.ts [hunk 1]
```diff
@@ -43,8 +43,8 @@ export const PropertyType = {
   ARRAY: 'array' as PropertyType,
 };
 
-export interface Property {
-  name: string;
+export interface Property<T = string> {
+  name: T;
```
Comment (vogievetsky) [inline src/base-immutable/base-immutable.ts:47]: what is the point of this?

Comment (lorem--ipsum) [reply src/base-immutable/base-immutable.ts:47]: To allow for stricter typing (i.e. `T` should be `keyof ValueType` but I'm having trouble accessing instance types from static properties — it's a known TypeScript limitations and I'm trying to find a better way)...

**Comment by vogievetsky** [comment]
gr 🎱

**Pull request merged**